src/odemis/driver/zeiss.py

```diff
@@ -0,0 +1,1010 @@
+            self._parseMessage(m)  # will update _output_buf
+
+        self._input_buf = msgs[-1]
+
+    def read(self, size=1):
+        ret = self._output_buf[:size]
+        self._output_buf = self._output_buf[len(ret):]
+
+        if len(ret) < size:
+            # simulate timeout
+            time.sleep(self.timeout)
+        return ret
+
+    def flush(self):
+        pass
+
+    def flushInput(self):
+        self._output_buf = ""
+
+    def close(self):
+        # using read or write will fail after that
+        del self._output_buf
+        del self._input_buf
+
+    def _sendAck(self, status):
+        self._output_buf += "%s\r\n" % (status,)
+
+    def _sendAnswer(self, status, ans):
+        self._output_buf += "%s%s\r\n" % (status, ans)
+
+    def _parseMessage(self, msg):
+        """
+        msg (str): the message to parse (without the \r)
+        return None: self._output_buf is updated if necessary
+        """
+
+        logging.debug("SIM: parsing %s", msg)
+        l = re.split(" ", msg)
+
+        com = l[0]
+        args = l[1:]
+        logging.debug("SIM: decoded message as %s %s", com, args)
+
+        # decode the command
+        if com == "VER?":
+            self._sendAck(RS_VALID)
+            self._sendAnswer(RS_SUCCESS, "SmartSEM Remote Control V01.23, DELMIC Sim")
+        elif com == "STG?":
+            self._sendAck(RS_VALID)
+            if self.waiting:
+                self._sendAnswer(RS_SUCCESS, "%s %s %s 1.0" % (self.posX, self.posY, self.posZ))
+            else:
+                self._sendAnswer(RS_SUCCESS, "%s %s %s 0.0" % (self.posX, self.posY, self.posZ))
+        elif com == "BBL?":
+            self._sendAck(RS_VALID)
+            self._sendAnswer(RS_SUCCESS, "%s" % self.blanker)
+        elif com == "MAG?":
+            self._sendAck(RS_VALID)
+            self._sendAnswer(RS_SUCCESS, "%s" % self.magnification)
+        elif com == "EXS?":
+            self._sendAck(RS_VALID)
+            self._sendAnswer(RS_SUCCESS, "%s" % self.external)
+        elif com == "PIX?":
+            self._sendAck(RS_VALID)
+            self._sendAnswer(RS_SUCCESS, "%s" % self.pixelSize)
+        elif com == "FOC?":
+            self._sendAck(RS_VALID)
+            self._sendAnswer(RS_SUCCESS, "%s" % self.focus)
+        elif com == "STG":
+            self._sendAck(RS_VALID)
+            self._sendAck(RS_SUCCESS)
+            self.target_posX, self.target_posY, self.target_posZ = [float(i) for i in args]
+            self._start_move = time.time()
+            dist = numpy.sqrt((self.posX - float(args[0])) ** 2 \
+                            + (self.posY - float(args[1])) ** 2 \
+                            + (self.posZ - float(args[2])) ** 2)
+            dur = dist / self._speed
+            self._end_move = time.time() + dur
+            self.dur = dur
+
+            self._stage_stop.clear()
+            self._mover = threading.Thread(target=self._run_move)
+            self._mover.daemon = True
```

Comment:
Probably no need to make it a daemon, as it should be a very short lived thread.